utils/src/metrics/histogram.rs

```diff
@@ -0,0 +1,53 @@
+use prometheus_client::metrics::histogram::Histogram;
+use std::time::Instant;
+
+/// Trait providing convenience methods for `Histogram`.
+pub trait HistogramExt {
+    fn guard(&self) -> HistogramGuard<'_>;
+}
+
+impl HistogramExt for Histogram {
+    fn guard(&self) -> HistogramGuard<'_> {
+        HistogramGuard {
+            histogram: self,
+            start: Instant::now(),
+            recorded: false,
+        }
+    }
+}
+
+/// Records the duration (in seconds) of a scope.
+///
+/// The duration is recorded when the guard is dropped, unless canceled or manually recorded.
+pub struct HistogramGuard<'a> {
+    start: Instant,
+    histogram: &'a Histogram,
+    recorded: bool,
+}
+
+impl HistogramGuard<'_> {
+    /// Manually record and prevent automatic recording
+    pub fn observe(&mut self) {
+        self.record();
+    }
+
+    /// Cancel automatic recording
+    pub fn cancel(&mut self) {
+        self.recorded = true;
+    }
+
+    /// Records the duration
+    fn record(&mut self) {
+        let duration = self.start.elapsed().as_secs_f64();
```

Comment:
I recommend using `ms` instead (often `s` is far too coarse for the ops this is used for)

Reply:
> When measuring elapsed time in seconds with a float64, you can accurately represent nanosecond-level differences for times up to a few months. After that, you'll start losing precision at the nanosecond level.

I'm fine to change to `ms` if we measure everything in `ms` (for consistency!), but if it actually uses f64 then measuring in seconds should be alright for precision down to the nanosecond level.

Reply:
Will leave as seconds per offline discussion